File: apps/apple-tv/MediaPlayer/Models/Progress.swift
import Foundation

struct WatchProgress: Codable, Identifiable {
    let id: Int64?
    let userId: Int64?
    let mediaId: Int64
    let mediaType: String
    let position: Int
    let duration: Int
    let completed: Bool
    let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, position, duration, completed
        case userId = "user_id"
        case mediaId = "media_id"
        case mediaType = "media_type"
        case updatedAt = "updated_at"
    }

    var progressPercentage: Double {
        guard duration > 0 else { return 0 }
        return Double(position) / Double(duration)
    }

    var formattedPosition: String {
        formatTime(position)
    }

    var formattedRemaining: String {
        let remaining = duration - position
        return formatTime(remaining)
    }

    private func formatTime(_ seconds: Int) -> String {
        let hours = seconds / 3600
        let minutes = (seconds % 3600) / 60
        let secs = seconds % 60

        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, secs)
        }
        return String(format: "%d:%02d", minutes, secs)
    }
}

struct ContinueWatchingItem: Codable, Identifiable {
    let media: Media
    let progress: WatchProgress

    var id: Int64 { media.id }
}

struct ContinueWatchingResponse: Codable {
    let items: [ContinueWatchingItem]
}

struct UpdateProgressRequest: Codable {
    let position: Int
    let duration: Int
    let mediaType: String
    let completed: Bool

    enum CodingKeys: String, CodingKey {
        case position, duration, completed
        case mediaType = "media_type"
    }
}

File: internal/api/handlers/progress.go
package handlers

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/stephencjuliano/media-server/internal/db"
)

type ProgressHandler struct {
	db *db.DB
}

func NewProgressHandler(database *db.DB) *ProgressHandler {
	return &ProgressHandler{db: database}
}

type UpdateProgressRequest struct {
	Position  int    `json:"position" binding:"required,min=0"`
	Duration  int    `json:"duration" binding:"required,min=0"`
	MediaType string `json:"media_type" binding:"required,oneof=movie tvshow episode"`
	Completed bool   `json:"completed"`
}

type ContinueWatchingItem struct {
	Media    *db.Media         `json:"media"`
	Progress *db.WatchProgress `json:"progress"`
}

// GetProgress returns the watch progress for a media item
func (h *ProgressHandler) GetProgress(c *gin.Context) {
	userID, _ := c.Get("user_id")
	mediaIDStr := c.Param("mediaId")
	mediaType := c.Query("type")

	if mediaType == "" {
		mediaType = "movie"
	}

	mediaID, err := strconv.ParseInt(mediaIDStr, 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid media ID"})
		return
	}

	progress, err := h.db.GetWatchProgress(userID.(int64), mediaID, db.MediaType(mediaType))
	if err == db.ErrNotFound {
		// Return empty progress
		c.JSON(http.StatusOK, gin.H{
			"media_id":  mediaID,
			"position":  0,
			"duration":  0,
			"completed": false,
		})
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch progress"})
		return
	}

	c.JSON(http.StatusOK, progress)
}

// UpdateProgress updates the watch progress for a media item
func (h *ProgressHandler) UpdateProgress(c *gin.Context) {
	userID, _ := c.Get("user_id")
	mediaIDStr := c.Param("mediaId")

	mediaID, err := strconv.ParseInt(mediaIDStr, 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid media ID"})
		return
	}

	var req UpdateProgressRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// Auto-mark as completed if near the end (95%)
	completed := req.Completed
	if req.Duration > 0 && float64(req.Position)/float64(req.Duration) > 0.95 {
		completed = true
	}

	err = h.db.UpsertWatchProgress(
		userID.(int64),
		mediaID,
		db.MediaType(req.MediaType),
		req.Position,
		req.Duration,
		completed,
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update progress"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"media_id":  mediaID,
		"position":  req.Position,
		"duration":  req.Duration,
		"completed": completed,
	})
}

// GetContinueWatching returns in-progress media for the current user
func (h *ProgressHandler) GetContinueWatching(c *gin.Context) {
	userID, _ := c.Get("user_id")
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "10"))

	if limit > 20 {
		limit = 20
	}

	progressItems, err := h.db.GetContinueWatching(userID.(int64), limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch continue watching"})
		return
	}

	// Fetch media details for each progress item
	var items []ContinueWatchingItem
	for _, p := range progressItems {
		media, err := h.db.GetMediaByID(p.MediaID)
		if err != nil {
			continue
		}
		items = append(items, ContinueWatchingItem{
			Media:    media,
			Progress: p,
		})
	}

	c.JSON(http.StatusOK, gin.H{"items": items})
}
